### DISCUSSION TITLE
All opportunities to reduce firmware size on ESP32

### DISCUSSION THREAD

**Discussion by AmirHmZz** [micropython/micropython#14473] [ESP32]
It would be nice to have a list of all opportunities to reduce firmware size to a minimum.

Here's some config I've tried already (from `sdkconfig.d2wd`):
```
CONFIG_COMPILER_OPTIMIZATION_SIZE=y
CONFIG_COMPILER_OPTIMIZATION_PERF=n
CONFIG_COMPILER_OPTIMIZATION_CHECKS_SILENT=y

CONFIG_LOG_MAXIMUM_LEVEL_ERROR=y
CONFIG_LOG_MAXIMUM_LEVEL_INFO=n

CONFIG_ETH_USE_SPI_ETHERNET=n
```

Is there anyway to disable these features:

- Bluetooth
- TLS
- lwIP
- micropython compiler

**Comment by felixdoerre** [comment]
You didn't specify which firmware variant you are using. For most of those features there are compiler flags, and they are disabled in some of the boards:
* bluetooth: `MICROPY_PY_BLUETOOTH`, see e.g. this board, which has it disabled:
https://github.com/micropython/micropython/blob/master/ports/esp32/boards/LOLIN_S2_MINI/mpconfigboard.h#L5
* TLS: `MICROPY_PY_SSL` and `MICROPY_SSL_MBEDTLS`
* lwip: this is not provided by micropython but instead used from the esp-idf. Micropython just implements a python interface to it: https://github.com/micropython/micropython/blob/fa23e4b093f81f03a24187c7ea0c928a9b4a661b/ports/esp32/modsocket.c. I don't see a macro to disable that interface, so maybe there is a flag missing there.
* micropython compiler: `MICROPY_ENABLE_COMPILER`

if you want to adjust `ESP32_GENERIC`, you could add those flags here: https://github.com/micropython/micropython/blob/master/ports/esp32/boards/ESP32_GENERIC/mpconfigboard.h

**Reply by AmirHmZz** [reply]
I've tried to disable TLS by setting `MICROPY_PY_SSL` and `MICROPY_SSL_MBEDTLS` to 0
but it seems cryptolib and hashlib should be excluded too.

**Reply by felixdoerre** [reply]
For that there is `MICROPY_PY_CRYPTOLIB` and `MICROPY_PY_HASHLIB`

**Reply by AmirHmZz** [reply]
Done. Now I'm struggling disabling `MICROPY_PY_BTREE` but it is throwing errors on build time.

**Comment by jonnor** [comment]
What is your current image size, and what is your target?

**Reply by AmirHmZz** [reply]
it is currently `1.26MB`. It would be great to reduce it under `1MB`.

What I've already tried:
```
MICROPY_OPT_COMPUTED_GOTO=0
MICROPY_PY_NETWORK_LAN=0
MICROPY_PY_BLUETOOTH=0
MICROPY_PY_SSL=0
MICROPY_SSL_MBEDTLS=0
MICROPY_PY_CRYPTOLIB=0
MICROPY_PY_HASHLIB=0
MICROPY_PY_WEBSOCKET=0
MICROPY_PY_WEBREPL=0
MICROPY_PY_MACHINE_I2S=0
```

**Comment by shariltumin** [comment]
You can try to reduce the number of frozen modules in your firmware. Try using this "manifest.py" in your "ports/esp32/boards" directory.

```bash
freeze("$(PORT_DIR)/modules")
# --include("$(MPY_DIR)/extmod/asyncio")

# Useful networking-related packages.
# --require("bundle-networking")
require("mip")
require("ntptime")
# --require("ssl")
require("requests")
# --require("webrepl")

# Provide urequests (which just forwards to requests) for backwards
# compatibility.
# --require("urequests")

# Require some micropython-lib modules.
# --require("aioespnow")
require("dht")
require("ds18x20")
require("neopixel")
require("onewire")
require("umqtt.robust")
require("umqtt.simple")
# --require("upysh")
```

It is up to you to choose which modules to include or exclude.

**Comment by xyzzy42** [comment]
It's possible to turn PPP off, https://github.com/micropython/micropython/pull/13640

And here's some more things you can turn off:
```
#define MICROPY_HW_ENABLE_SDCARD            (0)
#define MICROPY_PY_MACHINE_DAC              (0)
#define MICROPY_PY_MACHINE_I2S              (0)
#define MICROPY_PY_ESPNOW                   (0)
#define MICROPY_PY_NETWORK_LAN              (0)
```

ESP-IDF is configurable and things can be turned off there too.  Try CONFIG_ETH_USE_SPI_ETHERNET and CONFIG_LWIP_PPP_SUPPORT.  Also flash chip types you aren't using.
